pkgs/build-support/dotnet/make-nuget-deps/default.nix

```diff
@@ -52,7 +52,26 @@ lib.makeOverridable(
             rm -rf .signature.p7s
           '';
 
+          preInstall = ''
+            function traverseRename () {
+              for e in *
+              do
+                t="$(echo "$e" | sed -e "s/%20/\ /g" -e "s/%2B/+/g")"
+                [ "$t" != "$e" ] && mv -vn "$e" "$t"
+                if [ -d "$t" ]
+                then
+                  cd "$t"
+                  traverseRename
+                  cd ..
+                fi
+              done
+            }
+
+            traverseRename
+          '';
           installPhase = ''
+            runHook preInstall
+
```

Comment:
Copied from https://github.com/NixOS/nixpkgs/blob/89f8fd07fb125f182ad5652eff777940e09aa5c2/pkgs/build-support/dotnet/fetchnuget/default.nix#L30-L46

Is there a reason there's two different fetchNuGet implementations that don't behave the same?

Reply:
The other one is a related to dotnet-packages, which is more or less obsolete.

Which package requires this?  I'd like to make sure we solve this once and for all based on how the filenames are escaped.

Reply:
If I drop this change and rebuild I get:
```
CSC : error CS0006: Metadata file '/nix/store/c6728sp915jhl489f9k91jxhvcabh1sc-WiseUnpacker-1.3.3-nuget-deps/share/nuget/packages/system.valuetuple/4.5.0/ref/portable-net40+sl4+win8+wp8/System.ValueTuple.dll' could not be found [/build/source/WiseUnpacker/WiseUnpacker.csproj::TargetFramework=net40]
CSC : error CS0006: Metadata file '/nix/store/c6728sp915jhl489f9k91jxhvcabh1sc-WiseUnpacker-1.3.3-nuget-deps/share/nuget/packages/system.valuetuple/4.5.0/ref/portable-net40+sl4+win8+wp8/System.ValueTuple.dll' could not be found [/build/source/WiseUnpacker/WiseUnpacker.csproj::TargetFramework=net452]
```

When I look in the folder the right path shows up it's just URL encoded:
```
$ ls -l /nix/store/c6728sp915jhl489f9k91jxhvcabh1sc-WiseUnpacker-1.3.3-nuget-deps/share/nuget/packages/system.valuetuple/4.5.0/ref/
total 48
dr-xr-xr-x 2 root root 4096 Jan  1  1970 MonoAndroid10
dr-xr-xr-x 2 root root 4096 Jan  1  1970 MonoTouch10
dr-xr-xr-x 2 root root 4096 Jan  1  1970 net461
dr-xr-xr-x 2 root root 4096 Jan  1  1970 net47
dr-xr-xr-x 2 root root 4096 Jan  1  1970 netcoreapp2.0
dr-xr-xr-x 2 root root 4096 Jan  1  1970 netstandard2.0
dr-xr-xr-x 2 root root 4096 Jan  1  1970 portable-net40%2Bsl4%2Bwin8%2Bwp8
dr-xr-xr-x 2 root root 4096 Jan  1  1970 uap10.0.16299
dr-xr-xr-x 2 root root 4096 Jan  1  1970 xamarinios10
dr-xr-xr-x 2 root root 4096 Jan  1  1970 xamarinmac20
dr-xr-xr-x 2 root root 4096 Jan  1  1970 xamarintvos10
dr-xr-xr-x 2 root root 4096 Jan  1  1970 xamarinwatchos10
```

Reply:
Could we:

- put this in `unpackPhase`
- give it a more specific name like `unescapeNupkgPaths`
- make the replacement logic more robust

I'm thinking something like:

`(e=${e/\\/\\\\} && echo -e ${e//%/\\x})`

It's only really old versions of the nuget client (before https://github.com/NuGet/NuGet.Client/pull/1855) that escaped the files like this, but the client still supports unescaping them, so we should too, when pulling binary packages.  We also unpack .nupkg files in other places, but they will all be packed with recent versions of the client.

Even with recent tools, filenames will be mangled if you intentionally put a file like `foo%2B` in a package.

Reply:
Sounds good, seems to work fine on my end, is there a convenient way to check if this breaks other packages?